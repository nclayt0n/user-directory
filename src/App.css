/* http://meyerweb.com/eric/tools/css/reset/ 
   v2.0 | 20110126
   License: none (public domain)
*/

html,
body,
div,
span,
applet,
object,
iframe,
h1,
h2,
h3,
h4,
h5,
h6,
p,
blockquote,
pre,
a,
abbr,
acronym,
address,
big,
cite,
code,
del,
dfn,
em,
img,
ins,
kbd,
q,
s,
samp,
small,
strike,
strong,
sub,
sup,
tt,
var,
b,
u,
i,
center,
dl,
dt,
dd,
ol,
ul,
li,
fieldset,
form,
label,
legend,
table,
caption,
tbody,
tfoot,
thead,
tr,
th,
td,
article,
aside,
canvas,
details,
embed,
figure,
figcaption,
footer,
header,
hgroup,
menu,
nav,
output,
ruby,
section,
summary,
time,
mark,
audio,
video {
    margin: 0;
    padding: 0;
    border: 0;
    font-size: 100%;
    font: inherit;
    vertical-align: baseline;
}


/* HTML5 display-role reset for older browsers */

article,
aside,
details,
figcaption,
figure,
footer,
header,
hgroup,
menu,
nav,
section {
    display: block;
}

body {
    line-height: 1;
}

ol,
ul {
    list-style: none;
}

blockquote,
q {
    quotes: none;
}

blockquote:before,
blockquote:after,
q:before,
q:after {
    content: '';
    content: none;
}

table {
    border-collapse: collapse;
    border-spacing: 0;
}


/*User Directory css*/

:root {
    --oliveGreen: #4B443C;
    --burgundy: #4D0011;
    --gunmetalGray: #53565A;
    --coolGray: #E4E5E8;
}

* {
    border-radius: 6px;
    background-color: var(--gunmetalGray);
    letter-spacing: 1px;
    font-family: Arial, Helvetica, sans-serif;
}

.notFound {
    text-align: center;
    margin: 50px;
    padding: 20px;
    background-color: var(--burgundy);
    color: white;
}

.notFound p {
    text-align: center;
    margin: auto;
    background-color: var(--burgundy);
    color: white;
    font-size: 18px;
}

.landingPage {
    background-color: var(--burgundy);
    margin: 20px;
    padding: 40px;
    border: 2px solid var(--oliveGreen);
    font-family: Arial, Helvetica, sans-serif;
}

h1 {
    font-size: 33px;
    font-weight: bolder;
    text-align: center;
    background-color: var(--burgundy);
    color: white;
    margin: 50px auto;
    text-shadow: .5px .5px var(--oliveGreen);
    letter-spacing: 3px;
}

#browseButton {
    text-align: center;
    background-color: var(--coolGray);
    padding: 15px;
    text-shadow: .5px .5px var(--coolGray);
    letter-spacing: 2.5px;
    margin: auto;
    width: 80%;
}

.landingPage a {
    background-color: var(--coolGray);
    color: var(--burgundy);
    font-size: 26px;
    font-weight: bolder;
    text-decoration: none;
    text-shadow: .5px .5px var(--oliveGreen);
    cursor: pointer;
}

h2 {
    font-size: 28px;
    font-weight: bolder;
    text-align: center;
    background-color: var(--burgundy);
    color: white;
    margin: 10px;
    text-shadow: .5px .5px var(--oliveGreen);
    letter-spacing: 2.5px;
}

.userList {
    margin: 20px auto;
    border: 1px solid var(--oliveGreen);
    padding: 15px;
    width: 300px;
    background-color: var(--burgundy);
}

.userCard {
    border: 2px solid var(--gunmetalGrays);
    margin: 20px auto;
    padding: 5px;
    display: flex;
    background-color: var(--coolGray);
}

.userCard ul {
    flex: 2;
    height: fit-content;
    width: fit-content;
    background-color: var(--coolGray);
    padding: 5px;
    text-align: center
}

.userCard li {
    background-color: var(--coolGray);
    width: fit-content;
    margin: 5px;
    font-size: 18px;
    text-align: left;
    color: var(--burgundy);
    word-break: break-word;
}

.userName {
    font-weight: bold;
    font-size: 22px;
    text-shadow: .5px .5px var(--oliveGreen);
    margin: auto;
    cursor: pointer;
}

.userNameButton {
    font-weight: bold;
    font-size: 28px;
    text-shadow: .5px .5px var(--oliveGreen);
    margin: 5px auto;
    border: none;
    background-color: var(--coolGray);
    cursor: pointer;
    color: var(--burgundy);
}

.thumbnail {
    border: 2px solid var(--gunmetalGray);
    height: 55px;
    float: right;
    margin: auto;
}

.thumbnailContainer {
    background-color: var(--coolGray);
    margin: auto;
}

#currentPage {
    margin: 3px;
    background-color: var(--oliveGreen);
    color: white;
    text-shadow: .25px .25px var(--burgundy);
    width: 55px;
    padding: 5px;
    font-size: 18px;
    font-weight: bold;
}

.errorContainer {
    background-color: white;
    margin: 50px;
    padding: 30px;
    font-size: 20px;
    text-align: center;
    color: red;
    font-weight: bolder;
    text-shadow: .5px .25px var(--gunmetalGray);
    text-transform: uppercase;
}

.buttonContainer {
    width: fit-content;
    margin: 15px auto;
    display: flex;
    padding: 5px;
    background-color: var(--burgundy)
}

.button {
    border: 2px solid var(--burgundy);
    margin: 2px;
    background-color: var(--coolGray);
    color: var(--burgundy);
    width: 54px;
    padding: 5px;
    font-size: 16px;
    font-weight: bold;
    cursor: pointer;
}

#exportButton {
    width: fit-content;
    padding: 10px;
    text-decoration: none;
}

@media screen and (min-width:761px) {
    h2 {
        font-size: 36px;
    }
    .userList {
        width: 450px;
    }
    .userCard {
        width: 90%;
    }
    .userCard ul {
        flex: 2;
        margin: 10px;
        height: fit-content;
        width: fit-content;
    }
    .userName {
        font-weight: bold;
        font-size: 24px;
    }
    .mediumImg {
        border: 2px solid var(--gunmetalGray);
        height: 80px;
        margin: 10px;
    }
    .mediumImgContainer {
        background-color: var(--coolGray);
    }
    .landingPage {
        margin: 50px auto;
        width: 450px;
    }
    #beginLinkContainer {
        margin: auto;
        width: 70%;
    }
    .errorContainer {
        font-size: 26px;
    }
}

@media screen and (min-width:1024px) {
    .userCard ul {
        flex: 2;
    }
    .userCard li {
        font-size: 22px;
    }
    .userList {
        width: 650px;
    }
    .userCard {
        width: 90%;
    }
    .userCard ul {
        flex: 2;
        margin: 10px;
        height: fit-content;
        width: fit-content;
    }
    #userName {
        font-weight: bold;
        font-size: 28px;
    }
    h1 {
        font-size: 44px;
    }
    h2 {
        font-size: 44px;
    }
    .landingPage {
        margin: 50px auto;
        padding: 50px;
        width: 650px;
    }
    #beginLinkContainer {
        margin: auto;
        width: 50%;
    }
    .LargeImg {
        height: 128px;
        margin: 20px;
    }
    .button,
    #currentPage {
        width: 70px;
        font-size: 22px;
        font-weight: bold;
    }
    .notFound p {
        font-size: 22px;
    }
}