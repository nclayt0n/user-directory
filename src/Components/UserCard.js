import React from 'react';
import MediaQuery from 'react-responsive';

class UserCard extends React.Component{
   static defaultProps={
      age: 53,
      email: "louane.vidal@example.com",
      gender: "female",
      name: {title: "Miss", first: "Louane", last: "Vidal"},
      phone: "02-62-35-18-98",
      picture:{
      large: "https://randomuser.me/api/portraits/women/88.jpg",
      medium: "https://randomuser.me/api/portraits/med/women/88.jpg",
      thumbnail: "https://randomuser.me/api/portraits/thumb/women/88.jpg"}
   };
   state={hidden:true}
   render(){
      return(
         <section className='userCard'>
            {this.state.hidden===true
              ?<button  
                  aria-label="user's first and last name"
                  type='button' 
                  onClick={()=>{this.setState({hidden:false})}} 
                  className='userNameButton'>
                  {this.props.name.first} {this.props.name.last}
               </button>
               :null}
            {this.state.hidden===false
               ?<ul className={this.state.ulClassName}>
                  <li 
                     aria-label="user's first and last name"
                     className='userName' 
                     onClick={()=>{this.setState({hidden:true})}}>
                     {this.props.name.first} {this.props.name.last}
                  </li>
                  <MediaQuery maxWidth={760}>
                     <div className='thumbnailContainer'>
                        <img 
                           src={this.props.picture.thumbnail} 
                           alt='headshot of user' 
                           className='thumbnail'/>
                     </div>
                  </MediaQuery>
                  <li aria-label="user's age">{this.props.age}</li>
                  <li aria-label="user's gender">{this.props.gender}</li> 
                  <li aria-label="user's phone number">{this.props.phone}</li>
                  <li aria-label="user's email address">{this.props.email}</li>  
               </ul>
               :null}
               <MediaQuery minWidth={761} maxWidth={1023}>
                  <div className='mediumImgContainer'>
                     <img 
                        src={this.props.picture.medium} 
                        className='mediumImg' 
                        alt='headshot of user'/>
                  </div>
               </MediaQuery>
               <MediaQuery minWidth={1024}>
                  <img 
                     src={this.props.picture.large} 
                     className='LargeImg' 
                     alt='headshot of user'/>
               </MediaQuery>
         </section>
      )  
   }
}
export default UserCard