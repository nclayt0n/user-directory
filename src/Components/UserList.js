import React from 'react';
import UserCard from './UserCard';
import '../App.css';
import ApiService from '../services/api-service';
import ValidationError from '../Validation/ValidationError';
import {CSVLink} from 'react-csv';
const uuidv4 = require('uuid/v4');

class UserList extends React.Component{
    constructor(){
    super()
        this.state={
            page:1,
            pageLimit:10,
            users:[],
            error:''
        };
    }
    callApi=(num)=>{
        this.setState({page:num})
        ApiService.getUser(num)
        .then(([users]) => {
          this.setState({ users: users.results });
      })
      .catch(error => {
          this.setState({ error:error.message });
      })
    }
    componentDidMount(){
          ApiService.getUser(this.state.page)
          .then(([users]) => {
            this.setState({ users: users.results });
        })
        .catch(error => {
            this.setState({ error:error.message });
        });
    }
    render(){
        const headers = [
            { label: "First Name", key: "first_name" },
            { label: "Last Name", key: "last_name" }, 
            { label: "Age", key: "age" },
            { label: "Phone", key: "phone" },
            { label: "Email", key: "email" },
            { label: "Gender", key: "gender" },
            { label: "Thumbnail Picture", key: "thumbnail_picture" },
            { label: "Medium Picture", key: "medium_picture" },
            { label: "Large Picture", key: "large_picture" }
          ];
        const data=this.state.users.map(x=>({
            first_name:x.name.first,
            last_name:x.name.last, 
            age:x.dob.age,
            phone:x.phone,
            email:x.email,
            gender:x.gender,
            large_picture:x.picture.large,
            medium_picture:x.picture.medium,
            thumbnail_picture:x.picture.thumbnail
            }));
        let usersDisplayed=this.state.users.slice(0,10);
      return(
        <main className='userList'>
            <h2>User Directory</h2>
            <ValidationError errorMessage={this.state.error}/>
            {usersDisplayed.map((user)=>
                <UserCard
                    key={uuidv4()}
                    age={user.dob.age}
                    email={user.email}
                    gender={user.gender}
                    name={user.name}
                    phone={user.phone}
                    picture={user.picture}
                />
            )}
            <div className='buttonContainer'>
                {this.state.page===1 ? null: 
                    <button 
                        aria-label='button for previous page of users'
                        className='button' 
                        type='button' 
                        onClick={()=>this.callApi(this.state.page-1)}>Back
                    </button>}
                {this.state.page===1 ? null: 
                    <button 
                        aria-label='button for previous page of users'
                        className='button' 
                        type='button' 
                        onClick={()=>this.callApi(this.state.page-1)}>{this.state.page-1}
                    </button>}

                    <button 
                    aria-label='current page of users'
                        className='button' 
                        id='currentPage' 
                        type='button'>{this.state.page}
                    </button>
                {this.state.page===1 ? null: 
                    <button 
                        aria-label='button for next page of users'
                        className='button' 
                        type='button' 
                        onClick={()=>this.callApi(this.state.page+1)}>{this.state.page+1}
                    </button>}

                    <button 
                        aria-label='button for next page of users'
                        className='button' 
                        type='button' 
                        onClick={()=>this.callApi(this.state.page+1)}>Next
                    </button>
            </div>
            <div className='buttonContainer'>
                <CSVLink 
                    aria-label='button for exporting current users on page to CSV file'
                    className='button' 
                    id='exportButton' 
                    data={data} 
                    headers={headers}>
                    Export Page {this.state.page} to CSV
                </CSVLink>
            </div>

        </main>
      )  
    }
}
export default UserList