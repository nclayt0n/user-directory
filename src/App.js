import React from 'react';
import {Switch,Route} from 'react-router-dom';
import LandingPage from './Components/LandingPage';
import UserList from './Components/UserList';
import NotFound from './Components/NotFound';

class App extends React.Component {
  render(){
  return (
    <div className="App">
      <Switch>
        <Route exact path='/' component={LandingPage}/>
        <Route path='/user-list' component={UserList}/>
        <NotFound/>
      </Switch>
    </div>
  );
}}

export default App;
