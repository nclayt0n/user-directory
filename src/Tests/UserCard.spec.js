import React from 'react';
import ReactDOM from 'react-dom';
import UserList from '../Components/UserList';
import UserCard from '../Components/UserCard';
import { shallow } from 'enzyme';
import {BrowserRouter} from 'react-router-dom';
import toJson from 'enzyme-to-json';
import App from '../App';
 
 describe('<UserCard />', () => {
    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(
        <BrowserRouter>
            <App>
                <UserList>
                    <UserCard/>
                </UserList>
            </App>
        </BrowserRouter>,div);
        ReactDOM.unmountComponentAtNode(div);
    });
    it('snapshot', () => {
        const wrapper=shallow(<UserCard/>)
        expect(toJson(wrapper))
        .toMatchSnapshot()
    });
})
