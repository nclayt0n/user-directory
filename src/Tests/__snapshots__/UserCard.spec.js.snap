// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserCard /> snapshot 1`] = `
<section
  className="userCard"
>
  <button
    aria-label="user's first and last name"
    className="userNameButton"
    onClick={[Function]}
    type="button"
  >
    Louane
     
    Vidal
  </button>
  <MediaQuery
    maxWidth={1023}
    minWidth={761}
  >
    <div
      className="mediumImgContainer"
    >
      <img
        alt="headshot of user"
        className="mediumImg"
        src="https://randomuser.me/api/portraits/med/women/88.jpg"
      />
    </div>
  </MediaQuery>
  <MediaQuery
    minWidth={1024}
  >
    <img
      alt="headshot of user"
      className="LargeImg"
      src="https://randomuser.me/api/portraits/women/88.jpg"
    />
  </MediaQuery>
</section>
`;
